corehq/motech/forms.py

```diff
@@ -56,15 +56,38 @@ class BaseConnectionSettingsFormSet(forms.BaseModelFormSet):
 
     def __init__(self, *args, domain, **kwargs):
         super().__init__(*args, **kwargs)
+        self.domain = domain
         self.form_kwargs['domain'] = domain  # Passed by ``FormView.get_form_kwargs()``
 
+    @property
+    def deleted_forms(self):
+        """
+        Returns a list of forms that have been marked for deletion.
+
+        Excludes forms of ConnectionSettings that are in use.
+        """
+        deleted_forms = super().deleted_forms
+        ids_in_use = get_connection_ids_in_use(self.domain)
+        return [f for f in deleted_forms
+                if not (f.is_bound and f.instance.id in ids_in_use)]
+
+
+def get_connection_ids_in_use(domain):
+    from corehq.motech.dhis2.dbaccessors import get_dataset_maps
+
+    dataset_maps = get_dataset_maps(domain)
+    # So far only DataSetMaps use ConnectionSettings. When more things
+    # do (like Repeaters), this must check those too.
+    return {m.connection_settings_id for m in dataset_maps
+            if m.connection_settings_id}
```

Comment:
An alternative approach here would be to use a `ForeignKey` to link the two (it sorta already does), then I think Django should be able to tell you whether any related models reference a given connection (and you could use `on_delete=PROTECT`)

Reply:
I agree, that is definitely the way to go, but first `DataSetMap` (and other models for configuring DHIS2 integration) will need to be migrated from Couch to Postgres. Django doesn't support using `ForeignKey` [across databases](https://docs.djangoproject.com/en/3.0/topics/db/multi-db/#cross-database-relations).

Reply:
Oh right.  Carry on